Adds automatic pain emoting

comma opened the pull request:
Emotes vary in frequency and type depending on level of pain.
It's riding on the pain messaging so there's a cooldown already.

comma committed:
Adds automatic pain emoting

Emotes vary in frequency and type depending on level of pain.
It's riding on the pain messaging so there's a cooldown already.

SolatK commented:
Maybe make it unique for different species? Screaming GAS is funny. And can you make species based emote sound system for carbon/human?

comma commented:
I guess it's doable, but would need suggestions on what emotes to use for each species.

comma committed:
Made pain emotes species specific

Currently all pain feeling species share same ones except GAS and lizerbs.
Added some premade audibles, made hiss audible instead of visible (wtf)

PurpleMartinJCK commented:
I'll come up with a list

MistakeNot4892 commented on code/modules/emotes/definitions/audible.dm:
```diff
@@ -169,3 +169,19 @@
+	key = "roar"
+	emote_message_3p = "USER roars!"
+
+/decl/emote/audible/roar
```
Should be /bellow

ZCaliber commented:
Hmmmm. Mayhaps make a list of default pain emotes, but also allow it to be customized in Character creation? Perhaps for another PR in the future.

PsiOmegaDelta commented on code/modules/species/species.dm:
```diff
@@ -790,4 +790,14 @@ The slots that you can use are found in items_clothing.dm and are the inventory
+	else if(pain_power >= 40)
+		return pick("groan", "grunt", "shiver", "moan")
+	else if(pain_power >= 10)
+		return pick("groan", "grunt")
```
Something that is a little more generic and always ensures that a valid emote key is returned would be:
```
/datum/species
    var/list/intense_pain_emotes = list(/decl/emote/audible/scream, etc., etc.)
    var/list/moderate_pain_emotes = list(/decl/emote/audible/grunt, etc., etc.)
    var/list/light_pain_emotes = list(/decl/emote/audible/grunt, etc., etc.)

/datum/species/nabber
    intense_pain_emotes = list(/decl/emote/audible/bug_hiss)
    moderate_pain_emotes = list(/decl/emote/audible/bug_hiss)
    light_pain_emotes = null

[etc.]

/datum/species/proc/get_pain_emote(var/pain_power)
    if(!(species_flags & SPECIES_FLAG_NO_PAIN))
        return
    var/pain_emote
    if(pain_power >= 70)
        pain_emote = safepick(intense_pain_emotes)
    else if(pain_power >= 40)
        pain_emote = safepick(moderate_pain_emotes)
    else if(pain_power >= 10)
        pain_emote = safepick(light_pain_emotes)

    if(pain_emote)
        var/decl/emote/E = decls_repository.get_decl(pain_emote)
        return E.key
```

PsiOmegaDelta replied:
Hmm.. unathi have a bit higher pain threshold. Lemme think on it for a second

PsiOmegaDelta replied:
```
/datum/species
    // Order matters, higher pain level should be higher up
    var/list/pain_emotes_with_pain_level = list
    (
        list(/decl/emote/audible/scream, etc., etc.) = 70,
        list(/decl/emote/audible/grunt) = 40,
        list(/decl/emote/audible/moan) = 10,
    )

/datum/species/nabber
    pain_emotes_with_pain_level = list
    (
        list(/decl/emote/audible/bug_hiss) = 40
    )

/datum/species/unathi
    pain_emotes_with_pain_level = list
    (
        list(/decl/emote/audible/howl) = 70,
        list(/decl/emote/audible/grunt) = 40,
        list(/decl/emote/audible/groan) = 20,
    )

/datum/species/proc/get_pain_emote(var/pain_power)
    if(!(species_flags & SPECIES_FLAG_NO_PAIN))
        return
    for(var/pain_emotes in pain_emotes_with_pain_level)
        var/pain_level = pain_emotes_with_pain_level[pain_emotes]
        if(pain_level >= pain_power)
            // This assumes that if a pain-level has been defined it also has a list of emotes to go with it
            var/decl/emote/E = decls_repository.get_decl(pick(pain_emotes))
            return E.key
```

comma replied:
Yeah, string keys are p bad, might as well do this

PsiOmegaDelta commented on code/modules/species/station/nabber.dm:
```diff
@@ -387,3 +390,7 @@
 		if(28 to 40)	. = -2
 		else			. = -4
+
+/datum/species/nabber/get_pain_emote(var/mob/living/carbon/human/H, var/pain_power)
```
Probably want to remove this override
